File: scripts/bench_vs_supervision.py
"""Benchmark counterpart to examples/bench_vs_supervision.rs in warden-vision.

Same three scenarios, same sizes, so numbers are directly comparable:
1. NMS over 10,000 random boxes.
2. ByteTrack over 500 frames x 30 detections.
3. Box + label annotation over 500 frames x 50 boxes on a 1920x1080 image.
"""

import time
import warnings

import numpy as np
import supervision as sv

warnings.filterwarnings("ignore", category=FutureWarning)


class Rng:
    """Same xorshift PRNG as the Rust side, so both benchmarks draw identical data."""

    def __init__(self, seed):
        self.state = seed

    def next_f32(self):
        x = self.state
        x ^= (x << 13) & 0xFFFFFFFFFFFFFFFF
        x ^= x >> 7
        x ^= (x << 17) & 0xFFFFFFFFFFFFFFFF
        self.state = x
        return (x >> 11) / float(1 << 53)


def bench_nms():
    rng = Rng(42)
    n = 10_000
    xyxy = np.zeros((n, 4), dtype=np.float32)
    confidence = np.zeros(n, dtype=np.float32)
    class_id = np.zeros(n, dtype=int)
    for i in range(n):
        x = rng.next_f32() * 1900.0
        y = rng.next_f32() * 1000.0
        w = 20.0 + rng.next_f32() * 80.0
        h = 20.0 + rng.next_f32() * 80.0
        xyxy[i] = [x, y, x + w, y + h]
        confidence[i] = rng.next_f32()
        class_id[i] = i % 10

    detections = sv.Detections(xyxy=xyxy, confidence=confidence, class_id=class_id)

    iters = 20
    start = time.perf_counter()
    kept = 0
    for _ in range(iters):
        kept = len(detections.with_nms(threshold=0.5, class_agnostic=False))
    elapsed = time.perf_counter() - start
    print(
        f"[nms]        n={n:5} iters={iters:3} kept={kept:5} "
        f"total={elapsed * 1000:8.2f}ms avg={elapsed / iters * 1000:8.2f}ms"
    )


def bench_bytetrack():
    rng = Rng(7)
    frames = 500
    per_frame = 30

    all_frames = []
    for f in range(frames):
        xyxy = np.zeros((per_frame, 4), dtype=np.float32)
        confidence = np.zeros(per_frame, dtype=np.float32)
        class_id = np.zeros(per_frame, dtype=int)
        for i in range(per_frame):
            base_x = i * 60.0 + f * 0.5
            base_y = i * 15.0
            jitter = rng.next_f32() * 4.0 - 2.0
            xyxy[i] = [
                base_x + jitter,
                base_y + jitter,
                base_x + 40.0 + jitter,
                base_y + 40.0 + jitter,
            ]
            confidence[i] = 0.5 + rng.next_f32() * 0.5
            class_id[i] = i % 5
        all_frames.append(sv.Detections(xyxy=xyxy, confidence=confidence, class_id=class_id))

    tracker = sv.ByteTrack(
        track_activation_threshold=0.25,
        lost_track_buffer=30,
        minimum_matching_threshold=0.8,
    )

    start = time.perf_counter()
    total_tracked = 0
    for frame in all_frames:
        total_tracked += len(tracker.update_with_detections(frame))
    elapsed = time.perf_counter() - start
    print(
        f"[bytetrack]  frames={frames:4} per_frame={per_frame:3} tracked={total_tracked:6} "
        f"total={elapsed * 1000:8.2f}ms fps={frames / elapsed:8.1f}"
    )


def bench_annotate():
    rng = Rng(99)
    frames = 500
    per_frame = 50
    w, h = 1920, 1080

    all_frames = []
    for _ in range(frames):
        xyxy = np.zeros((per_frame, 4), dtype=np.float32)
        confidence = np.zeros(per_frame, dtype=np.float32)
        class_id = np.zeros(per_frame, dtype=int)
        tracker_id = np.zeros(per_frame, dtype=int)
        for i in range(per_frame):
            x = rng.next_f32() * (w - 100.0)
            y = rng.next_f32() * (h - 100.0)
            xyxy[i] = [x, y, x + 80.0, y + 60.0]
            confidence[i] = 0.9
            class_id[i] = i % 10
            tracker_id[i] = i
        all_frames.append(
            sv.Detections(
                xyxy=xyxy,
                confidence=confidence,
                class_id=class_id,
                tracker_id=tracker_id,
            )
        )

    box_annotator = sv.BoxAnnotator(thickness=2)
    label_annotator = sv.LabelAnnotator(text_scale=0.5, text_thickness=1, text_padding=4)

    start = time.perf_counter()
    for dets in all_frames:
        frame = np.zeros((h, w, 3), dtype=np.uint8)
        frame = box_annotator.annotate(frame, dets)
        frame = label_annotator.annotate(frame, dets)
    elapsed = time.perf_counter() - start
    print(
        f"[annotate]   frames={frames:4} per_frame={per_frame:3} {w}x{h} "
        f"total={elapsed * 1000:8.2f}ms fps={frames / elapsed:8.1f}"
    )


if __name__ == "__main__":
    print(f"supervision {sv.__version__} benchmark")
    bench_nms()
    bench_bytetrack()
    bench_annotate()

File: examples/bench_vs_supervision.rs
//! Benchmark used to compare warden-vision against Python `supervision`.
//!
//! Three scenarios, sized to match `scripts/bench_vs_supervision.py`:
//! 1. NMS over 10,000 random boxes.
//! 2. ByteTrack over 500 frames x 30 detections.
//! 3. Box + label annotation over 500 frames x 50 boxes on a 1920x1080 image.
//!
//! Run with: `cargo run --release --example bench_vs_supervision --features annotate-image`

use std::time::Instant;
use warden_vision::annotators::{Annotator, BoxAnnotator, ColorPalette, LabelAnnotator};
use warden_vision::core::{Detection, Detections};
use warden_vision::geometry::Position;
use warden_vision::tracker::ByteTracker;

// Small xorshift PRNG so results are reproducible without pulling in `rand`.
struct Rng(u64);
impl Rng {
    fn next_f32(&mut self) -> f32 {
        self.0 ^= self.0 << 13;
        self.0 ^= self.0 >> 7;
        self.0 ^= self.0 << 17;
        (self.0 >> 11) as f32 / (1u64 << 53) as f32
    }
}

fn bench_nms() {
    let mut rng = Rng(42);
    let n = 10_000;
    let mut dets = Vec::with_capacity(n);
    for i in 0..n {
        let x = rng.next_f32() * 1900.0;
        let y = rng.next_f32() * 1000.0;
        let w = 20.0 + rng.next_f32() * 80.0;
        let h = 20.0 + rng.next_f32() * 80.0;
        dets.push(Detection::new([x, y, x + w, y + h], rng.next_f32(), i % 10));
    }
    let detections = Detections::new(dets);

    let iters = 20;
    let start = Instant::now();
    let mut kept = 0;
    for _ in 0..iters {
        kept = detections.non_max_suppression(0.5, false).len();
    }
    let elapsed = start.elapsed();
    println!(
        "[nms]        n={n:5} iters={iters:3} kept={kept:5} total={:8.2?} avg={:8.2?}",
        elapsed,
        elapsed / iters
    );

    #[cfg(feature = "parallel")]
    {
        let start = Instant::now();
        let mut kept = 0;
        for _ in 0..iters {
            kept = detections.non_max_suppression_parallel(0.5, false).len();
        }
        let elapsed = start.elapsed();
        println!(
            "[nms-par]    n={n:5} iters={iters:3} kept={kept:5} total={:8.2?} avg={:8.2?}",
            elapsed,
            elapsed / iters
        );
    }
}

fn bench_bytetrack() {
    let mut rng = Rng(7);
    let frames = 500;
    let per_frame = 30;

    // Precompute per-frame detections so tracker time is isolated.
    let mut all_frames = Vec::with_capacity(frames);
    for f in 0..frames {
        let mut dets = Vec::with_capacity(per_frame);
        for i in 0..per_frame {
            let base_x = (i as f32) * 60.0 + (f as f32) * 0.5;
            let base_y = (i as f32) * 15.0;
            let jitter = rng.next_f32() * 4.0 - 2.0;
            dets.push(Detection::new(
                [
                    base_x + jitter,
                    base_y + jitter,
                    base_x + 40.0 + jitter,
                    base_y + 40.0 + jitter,
                ],
                0.5 + rng.next_f32() * 0.5,
                i % 5,
            ));
        }
        all_frames.push(Detections::new(dets));
    }

    let mut tracker = ByteTracker::new(0.25, 30, 0.8, 0.5);
    let start = Instant::now();
    let mut total_tracked = 0usize;
    for frame in &all_frames {
        total_tracked += tracker.update(frame).len();
    }
    let elapsed = start.elapsed();
    println!(
        "[bytetrack]  frames={frames:4} per_frame={per_frame:3} tracked={total_tracked:6} total={:8.2?} fps={:8.1}",
        elapsed,
        frames as f64 / elapsed.as_secs_f64()
    );
}

fn bench_annotate() {
    let mut rng = Rng(99);
    let frames = 500;
    let per_frame = 50;
    let (w, h) = (1920u32, 1080u32);

    let mut all_frames = Vec::with_capacity(frames);
    for _ in 0..frames {
        let mut dets = Vec::with_capacity(per_frame);
        for i in 0..per_frame {
            let x = rng.next_f32() * (w as f32 - 100.0);
            let y = rng.next_f32() * (h as f32 - 100.0);
            let mut d = Detection::new([x, y, x + 80.0, y + 60.0], 0.9, i % 10);
            d.tracker_id = Some(i);
            dets.push(d);
        }
        all_frames.push(Detections::new(dets));
    }

    let box_annotator = BoxAnnotator::new(ColorPalette::default(), 2);
    let label_annotator = LabelAnnotator::new(ColorPalette::default(), Position::TopLeft, 1, 4.0);

    let start = Instant::now();
    for dets in &all_frames {
        let mut frame = image::RgbaImage::new(w, h);
        box_annotator.annotate(&mut frame, dets).unwrap();
        label_annotator.annotate(&mut frame, dets).unwrap();
        std::hint::black_box(&frame);
    }
    let elapsed = start.elapsed();
    println!(
        "[annotate]   frames={frames:4} per_frame={per_frame:3} {w}x{h} total={:8.2?} fps={:8.1}",
        elapsed,
        frames as f64 / elapsed.as_secs_f64()
    );
}

fn main() {
    println!("warden-vision benchmark ({})", std::env::consts::ARCH);
    bench_nms();
    bench_bytetrack();
    bench_annotate();
}
